pkg/instrument/sqlite.go

```diff
@@ -0,0 +1,68 @@
+package instrument
+
+import (
+	"time"
+	_ "unsafe"
+
+	"github.com/mattn/go-sqlite3"
+)
+
+// #cgo LDFLAGS: -lsqlite3
+// #include "extension.h"
+import "C"
+
+// StartSQLiteMonitoring instruments sqlite to collect metrics
+// on all connection opened after the call (i.e. existing connections
+// will not be instrumented).
+// Monitoring must be stopped by calling [StopSQLiteMonitoring].
+func StartSQLiteMonitoring() error {
+	if err := C.sqlite3_instrument(); err != sqlite3.SQLITE_OK {
+		return sqlite3.ErrNo(err)
+	}
+	return nil
+}
+
+// StopSQLiteMonitoring removes instrumentation from sqlite.
+// Existing connections will still be insturmented until they
+// are closed.
+func StopSQLiteMonitoring() { C.sqlite3_deinstrument() }
+
+type SQLiteMetrics struct {
+	// These fields follow the same order as the `struct sqlite3_metrics_s`.
+
+	PagesCacheWrite uint64
+	PagesCacheHit   uint64
+	PagesCacheMiss  uint64
+	PagesCacheSpill uint64
+
+	TransactionReadTime  time.Duration
+	TransactionWriteTime time.Duration
+}
+
+// ResetSQLiteMetrics returns the current metrics and
+// resets their value to 0.
```

Comment:
I think the model is slightly different to the C-side here, so there isn't anything external being reset to zero just here. Perhaps this function could be renamed? Perhaps something like `FinalizeSQLiteMetrics` (with the US spelling of 'finalise' :P)? I'm really not sure what a better name would be here

Reply:
This will affect subsequent calls though, so the reset is an effect which is visible from the outside and it should be valid. Moreover, while the *current* usage is clearly using it just to reset (and discards the result) that might not be the case during live server executions as you would use the result of this function. In that case wouldn't it be weird if I would take data multiple times from a `FinalizeXXX` function?